In [1]:
import pandas as pd

df = pd.read_csv('/kaggle/input/olympic-athelete-dataset/athlete_events.csv')

df.head(5)

ID                      Name Sex   Age  Height  Weight            Team  \
0   1                 A Dijiang   M  24.0   180.0    80.0           China   
1   2                  A Lamusi   M  23.0   170.0    60.0           China   
2   3       Gunnar Nielsen Aaby   M  24.0     NaN     NaN         Denmark   
3   4      Edgar Lindenau Aabye   M  34.0     NaN     NaN  Denmark/Sweden   
4   5  Christine Jacoba Aaftink   F  21.0   185.0    82.0     Netherlands   

   NOC        Games  Year  Season       City          Sport  \
0  CHN  1992 Summer  1992  Summer  Barcelona     Basketball   
1  CHN  2012 Summer  2012  Summer     London           Judo   
2  DEN  1920 Summer  1920  Summer  Antwerpen       Football   
3  DEN  1900 Summer  1900  Summer      Paris     Tug-Of-War   
4  NED  1988 Winter  1988  Winter    Calgary  Speed Skating   

                              Event Medal  
0       Basketball Men's Basketball   NaN  
1      Judo Men's Extra-Lightweight   NaN  
2           Football Men's Football   NaN  
3       Tug-Of-War Men's Tug-Of-War  Gold  
4  Speed Skating Women's 500 metres   NaN

In [2]:
df.shape

(271116, 15)

In [3]:
df.columns

Index(['ID', 'Name', 'Sex', 'Age', 'Height', 'Weight', 'Team', 'NOC', 'Games',
       'Year', 'Season', 'City', 'Sport', 'Event', 'Medal'],
      dtype='object')

In [4]:
df.describe()

ID            Age         Height         Weight  \
count  271116.000000  261642.000000  210945.000000  208241.000000   
mean    68248.954396      25.556898     175.338970      70.702393   
std     39022.286345       6.393561      10.518462      14.348020   
min         1.000000      10.000000     127.000000      25.000000   
25%     34643.000000      21.000000     168.000000      60.000000   
50%     68205.000000      24.000000     175.000000      70.000000   
75%    102097.250000      28.000000     183.000000      79.000000   
max    135571.000000      97.000000     226.000000     214.000000   

                Year  
count  271116.000000  
mean     1978.378480  
std        29.877632  
min      1896.000000  
25%      1960.000000  
50%      1988.000000  
75%      2002.000000  
max      2016.000000

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271116 entries, 0 to 271115
Data columns (total 15 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   ID      271116 non-null  int64  
 1   Name    271116 non-null  object 
 2   Sex     271116 non-null  object 
 3   Age     261642 non-null  float64
 4   Height  210945 non-null  float64
 5   Weight  208241 non-null  float64
 6   Team    271116 non-null  object 
 7   NOC     271116 non-null  object 
 8   Games   271116 non-null  object 
 9   Year    271116 non-null  int64  
 10  Season  271116 non-null  object 
 11  City    271116 non-null  object 
 12  Sport   271116 non-null  object 
 13  Event   271116 non-null  object 
 14  Medal   39783 non-null   object 
dtypes: float64(3), int64(2), object(10)
memory usage: 31.0+ MB


In [6]:
import numpy as np

df["Male"] = np.where(df['Sex'] == "M", 1, 0)
df["Female"] = np.where(df['Sex'] == "F", 1, 0)

df["Gold"] = np.where(df["Medal"] == 'Gold', 1, 0)
df["Silver"] = np.where(df["Medal"] == 'Silver', 1, 0)
df["Bronze"] = np.where(df["Medal"] == 'Bronze', 1, 0)
df["NoMedal"] = np.where(df["Medal"].isnull() == True, 1, 0)

df.drop(columns=['Medal'], inplace=True)

df.head(5)

ID                      Name Sex   Age  Height  Weight            Team  \
0   1                 A Dijiang   M  24.0   180.0    80.0           China   
1   2                  A Lamusi   M  23.0   170.0    60.0           China   
2   3       Gunnar Nielsen Aaby   M  24.0     NaN     NaN         Denmark   
3   4      Edgar Lindenau Aabye   M  34.0     NaN     NaN  Denmark/Sweden   
4   5  Christine Jacoba Aaftink   F  21.0   185.0    82.0     Netherlands   

   NOC        Games  Year  Season       City          Sport  \
0  CHN  1992 Summer  1992  Summer  Barcelona     Basketball   
1  CHN  2012 Summer  2012  Summer     London           Judo   
2  DEN  1920 Summer  1920  Summer  Antwerpen       Football   
3  DEN  1900 Summer  1900  Summer      Paris     Tug-Of-War   
4  NED  1988 Winter  1988  Winter    Calgary  Speed Skating   

                              Event  Male  Female  Gold  Silver  Bronze  \
0       Basketball Men's Basketball     1       0     0       0       0   
1      Judo Men's Extra-Lightweight     1       0     0       0       0   
2           Football Men's Football     1       0     0       0       0   
3       Tug-Of-War Men's Tug-Of-War     1       0     1       0       0   
4  Speed Skating Women's 500 metres     0       1     0       0       0   

   NoMedal  
0        1  
1        1  
2        1  
3        0  
4        1

# Printing the number of unique values in each row

In [7]:
def count_uniques(columnName):
    return len(df[columnName].unique())

columnNames = [column for column in df.columns]
for columnName in columnNames:
    print(f'{columnName} {count_uniques(columnName)}')

ID 135571
Name 134732
Sex 2
Age 75
Height 96
Weight 221
Team 1184
NOC 230
Games 51
Year 35
Season 2
City 42
Sport 66
Event 765
Male 2
Female 2
Gold 2
Silver 2
Bronze 2
NoMedal 2


In [8]:
df.isnull().sum()

ID             0
Name           0
Sex            0
Age         9474
Height     60171
Weight     62875
Team           0
NOC            0
Games          0
Year           0
Season         0
City           0
Sport          0
Event          0
Male           0
Female         0
Gold           0
Silver         0
Bronze         0
NoMedal        0
dtype: int64

In [9]:
df.isnull().sum() * 100 / df.shape[0]

ID          0.000000
Name        0.000000
Sex         0.000000
Age         3.494445
Height     22.193821
Weight     23.191180
Team        0.000000
NOC         0.000000
Games       0.000000
Year        0.000000
Season      0.000000
City        0.000000
Sport       0.000000
Event       0.000000
Male        0.000000
Female      0.000000
Gold        0.000000
Silver      0.000000
Bronze      0.000000
NoMedal     0.000000
dtype: float64

* We can use mean imputation for filling missing values in Age, as percent of missing values is 3.49% (< 5%)

* Since height and weight are having a substantial amount of missing values, we will handle it not by using simple imputaion methods, but by 

In [10]:
df_height_missing = df[df['Height'].isna()]
df_height_full = df[df['Height'].notna()]

df_weight_missing = df[df['Weight'].isna()]
df_weight_full = df[df['Weight'].notna()]

* On basic knowledge, we can say that height and weight should have high correlation with Sport, Sex and Age. So we will use those columns for prediction

In [11]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

preprocessor = ColumnTransformer(
    transformers=[
        ('cat', OneHotEncoder(handle_unknown='ignore'), ['Sex', 'Sport', 'Age'])
    ], remainder='passthrough'
)

In [12]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.pipeline import Pipeline

X_height = df_height_full[['Sex', 'Sport', 'Age']]
y_height = df_height_full['Height']

X_train, X_test, y_train, y_test = train_test_split(X_height, y_height, test_size=0.2, random_state=42)

height_pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('model', RandomForestRegressor(n_estimators=100, random_state=42))
])

height_pipeline.fit(X_train, y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('cat',
                                                  OneHotEncoder(handle_unknown='ignore'),
                                                  ['Sex', 'Sport', 'Age'])])),
                ('model', RandomForestRegressor(random_state=42))])

In [13]:
X_height_missing = df_height_missing[['Sex', 'Sport', 'Age']]
df_height_missing.loc[:, 'Height'] = height_pipeline.predict(X_height_missing)

df_height_missing.head(5)

ID                            Name Sex   Age      Height  Weight  \
2    3             Gunnar Nielsen Aaby   M  24.0  176.499261     NaN   
3    4            Edgar Lindenau Aabye   M  34.0  180.994103     NaN   
29  10  Einar Ferdinand "Einari" Aalto   M  26.0  186.397008     NaN   
35  15            Arvo Ossian Aaltonen   M  22.0  185.396331     NaN   
36  15            Arvo Ossian Aaltonen   M  22.0  185.396331     NaN   

              Team  NOC        Games  Year  Season       City       Sport  \
2          Denmark  DEN  1920 Summer  1920  Summer  Antwerpen    Football   
3   Denmark/Sweden  DEN  1900 Summer  1900  Summer      Paris  Tug-Of-War   
29         Finland  FIN  1952 Summer  1952  Summer   Helsinki    Swimming   
35         Finland  FIN  1912 Summer  1912  Summer  Stockholm    Swimming   
36         Finland  FIN  1912 Summer  1912  Summer  Stockholm    Swimming   

                                     Event  Male  Female  Gold  Silver  \
2                  Football Men's Football     1       0     0       0   
3              Tug-Of-War Men's Tug-Of-War     1       0     1       0   
29     Swimming Men's 400 metres Freestyle     1       0     0       0   
35  Swimming Men's 200 metres Breaststroke     1       0     0       0   
36  Swimming Men's 400 metres Breaststroke     1       0     0       0   

    Bronze  NoMedal  
2        0        1  
3        0        0  
29       0        1  
35       0        1  
36       0        1